koku/api/iam/view/user.py

```diff
@@ -190,3 +196,70 @@ def destroy(self, request, *args, **kwargs):
             HTTP/1.1 204 NO CONTENT
         """
         return super().destroy(request=request, args=args, kwargs=kwargs)
+
+    @action(methods=['put'], detail=True, permission_classes=[IsAuthenticated],
+            url_path='reset-password', url_name='reset-password',
+            serializer_class=serializers.PasswordChangeSerializer)
+    def reset_password(self, request, uuid=None):
+        """Reset a user's password.
+
+        @api {put} /api/v1/user/:id/reset-password/ Reset user password
+        @apiName ResetUserPassword
+        @apiGroup Users
+        @apiVersion 1.0.0
+        @apiDescription Reset a user's password.
+
+        @apiHeader {String} token User authorizaton token.
+        @apiHeaderExample {json} Header-Example:
+            {
+                "Authorizaton": "Token 45138a913da44ab89532bab0352ef84b"
+            }
+
+        @apiParam {Number} id User unique ID.
+
+        @apiParam (Request Body) {String} token The reset token for the user
+        @apiParam (Request Body) {String} password The password for the user
+        @apiParamExample {json} Request Body:
+            {
+            "token": "cc0111c4ffab478e802c-579bd14f5818",
+            "password": "str0ng!P@ss"
+            }
+
+        @apiSuccess {Number} id The identifier of the user.
+        @apiSuccess {String} username  The name of the user.
+        @apiSuccess {String} email  The email address of the user.
+        @apiSuccessExample {json} Success-Response:
+            HTTP/1.1 200 OK
+            {
+                "id": "57e60f90-8c0c-4bd1-87a0-2143759aae1c",
+                "username": "smithj",
+                "email": "smithj@mytechco.com"
+            }
+        """
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        user = get_object_or_404(model.User.objects.all(), uuid=uuid)
+
+        token = request.data.get('token')
+        password = request.data.get('password')
+        now = timezone.now()
+        # check for reset token
+        token_qs = model.ResetToken.objects.filter(token=token,
+                                                   user=user,
+                                                   used=False,
+                                                   expiration_date__gt=now)
+        if not token_qs.exists():
+            error = {
+                'token': [_('Token has expired or is invalid.')]
+            }
+            raise ValidationError(error)
+
+        reset_token = token_qs.first()
+        user.set_password(password)
```

Comment:
Might be nice to not allow their previous password to be provided as the current password.

Reply:
👍

Reply:
@adberglund Done.